framework/src/samplers/Sampler.C

```diff
@@ -12,103 +12,73 @@
-                    false),
+    SetupInterface(this),
     DistributionInterface(this),
-    Restartable(parameters, "Samplers"),
-    _tid(getParam<THREAD_ID>("_tid")),
-    _reseed_for_new_sample(getParam<bool>("reseed")),
-    _dist_names(getParam<std::vector<DistributionName>>("distributions")),
-    _var_names(getParam<std::vector<std::string>>("perturb_parameters")),
-    _current_sample(0),
-    _failed_runs(true)
+    _distribution_names(getParam<std::vector<DistributionName>>("distributions")),
+    _seed(getParam<unsigned int>("seed"))
 {
-  _var_dist_map.clear();
-  _probability_weight.clear();
-  _var_value_hist.clear();
-  if (_var_names.size() != _dist_names.size())
-    mooseError("The size of perturb_parameters (",
-               _var_names.size(),
-               ") != the size of distributions (",
-               _dist_names.size(),
-               ").");
-  if (!_var_names.empty())
-  {
-    for (unsigned int i = 0; i < _var_names.size(); ++i)
-    {
-      _var_dist_map[_var_names[i]] = &getDistributionByName(_dist_names[i]);
-      _var_value_map[_var_names[i]] = 0.0;
-      _var_value_hist[_var_names[i]] = std::vector<Real>();
-    }
-  }
+  for (const DistributionName & name : _distribution_names)
+    _distributions.push_back(&getDistributionByName(name));
+  setNumberOfRequiedRandomSeeds(1);
 }
 
 void
-Sampler::generateSamples()
+Sampler::execute()
 {
+  // Get the samples then save the state so that subsequent calls to getSamples returns the same
+  // random numbers until this execute command is called again.
+  getSamples();
+  _generator.saveState();
 }
 
-std::vector<std::string>
-Sampler::getSampledVariableNames()
+std::vector<DenseMatrix<Real>>
+Sampler::getSamples()
 {
-  return _var_names;
+  _generator.restoreState();
+  sampleSetUp();
+  std::vector<DenseMatrix<Real>> output = sample();
+  sampleTearDown();
+  return output;
 }
 
-std::vector<Real>
-Sampler::getSampledValues(const std::vector<std::string> & variableNames)
+double
+Sampler::rand(const unsigned int index)
 {
-  std::vector<Real> vals;
-  for (auto & var_name : variableNames)
-  {
-    Real val = getSampledValue(var_name);
-    vals.push_back(val);
-  }
-  return vals;
+  mooseAssert(index < _generator.size(), "The seed number index does not exists.");
+  return _generator.rand(index);
 }
 
-Real
-Sampler::getSampledValue(const std::string & variableName)
+void
+Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
 {
-  std::map<std::string, Real>::iterator it;
-  it = _var_value_map.find(variableName);
-  if (it == _var_value_map.end())
-    mooseError(
-        "Could not find the parameter: ", variableName, " in the list of perturbed parameters!");
-  return it->second;
-}
+  if (number == 0)
+    mooseError("The number of seeds must be larger than zero.");
 
-std::vector<Real>
-Sampler::getProbabilityWeights()
-{
-  return _probability_weight;
-}
+  for (std::size_t i = 0; i < number; ++i)
+    _generator.seed(i, _seed + i);
```

Comment:
This is bad idea. This almost guarantees that multiple simulations where seeds overlap will be invalid. seeds = {0, 1, **2, 3**} won't produce independent results from seeds {**2, 3**, 4, 5}. `RandomData` solved this problem with a master generator that produces seeds for slave generators.